packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart

```diff
@@ -0,0 +1,111 @@
+
+void main() {
+  group('Android migration', () {
+    
+    testWithoutContext('migrators succeed', () {
+      final FakeAndroidMigrator fakeAndroidMigrator = FakeAndroidMigrator();
+      final ProjectMigration migration = ProjectMigration(<ProjectMigrator>[fakeAndroidMigrator]);
+      migration.run();
+    });
+
+    group('migrate the Gradle "clean" task to lazy declaration', () {
+      late MemoryFileSystem memoryFileSystem;
+      late BufferLogger bufferLogger;
+      late FakeAndroidProject project;
+      late File topLevelGradleBuildFile;
+
+      setUp(() {
+        memoryFileSystem = MemoryFileSystem();
+        bufferLogger = BufferLogger.test();
+        project = FakeAndroidProject();
+        project.hostAppGradleRoot = memoryFileSystem.currentDirectory;
+        topLevelGradleBuildFile = project.hostAppGradleRoot.childFile('build.gradle');
+      });
+
+      testUsingContext('skipped if files are missing', () {
+        final TopLevelGradleBuildFileMigration androidProjectMigration = TopLevelGradleBuildFileMigration(
+          project,
+          bufferLogger,
+        );
+        androidProjectMigration.migrate();
+        expect(topLevelGradleBuildFile.existsSync(), isFalse);
+
+        expect(bufferLogger.traceText, contains('Top-level Gradle build file not found, skipping "clean" task migration.'));
+        expect(testLogger.statusText, isEmpty);
+      });
+
+      testUsingContext('skipped if nothing to upgrade', () {
+        topLevelGradleBuildFile.writeAsStringSync('''
+tasks.register("clean", Delete) {
+  delete rootProject.buildDir
+}
+        ''');
+
+        final TopLevelGradleBuildFileMigration androidProjectMigration = TopLevelGradleBuildFileMigration(
+          project,
+          bufferLogger,
+        );
+        final DateTime topLevelGradleBuildFileLastModified = topLevelGradleBuildFile.lastModifiedSync();
+        androidProjectMigration.migrate();
+
+        expect(topLevelGradleBuildFile.lastModifiedSync(), topLevelGradleBuildFileLastModified);
+        expect(testLogger.statusText, isEmpty);
+      });
+
+      testUsingContext('top-level build.gradle is migrated', () {
+        topLevelGradleBuildFile.writeAsStringSync('''
+task clean(type: Delete) {
+    delete rootProject.buildDir
+}
+''');
+
+        final TopLevelGradleBuildFileMigration androidProjectMigration = TopLevelGradleBuildFileMigration(
+          project,
+          bufferLogger,
+        );
+        androidProjectMigration.migrate();
+
+        expect(bufferLogger.traceText, contains('Migrating "clean" Gradle task to lazy declaration style.'));
+        expect(testLogger.statusText, isEmpty);
+
+        expect(topLevelGradleBuildFile.readAsStringSync(), equals('''
+tasks.register("clean", Delete) {
+    delete rootProject.buildDir
+}
+'''));
+      });
+    });
+  });
+}
+
+class FakeAndroidProject extends Fake implements AndroidProject {
+  @override
+  Directory hostAppGradleRoot = MemoryFileSystem.test().currentDirectory;
```

Comment:
This will be the root (aka `/`) of the simulated filesystem. Probably cleaner to put it a bit deeper, like an `android` subdirectory. Potentially in the future something else in this group of tests will want to put something else in the test filesystem.

Reply:
Yeah, I agree.

I rewrote `FakeAndroidProject` to be:

```dart
class FakeAndroidProject extends Fake implements AndroidProject {
  FakeAndroidProject(this.hostAppGradleRoot);

  @override
  Directory hostAppGradleRoot;
}
```

because the path was overridden anyway in `setUp()`.